### PR TITLE
Added dataset member to HTMLElement object.

### PR DESCRIPTION
**Pull request by Indomian**
This adds dataset field to HTMLElement object. Also there is DOMStringMap implementation, but it's not used until Proxy available for NodeJS without console argument.

Signed-off-by: Egor Bolgov <e.bolgov@semrush.com>

### CODE REVIEW

File: src/HTMLElement.js [hunk 1]
```diff
@@ -25,6 +26,13 @@ export default class HTMLElement extends Element {
          * @readonly
          */
         this.classList = new ClassList(this);
+        /**
+         * returns dataset object
+         * @type {DOMStringMap} dataset
+         * @readonly
+         */
+        this.dataset = {};
```
Comment (christophehurpeau) [inline src/HTMLElement.js]: prefer a variable named `_dataset` and a getter dataset

File: src/HTMLElement.js [hunk 2]
```diff
@@ -51,6 +59,10 @@ export default class HTMLElement extends Element {
         if (attributeName === 'class') {
             this.classList._parse(value || '');
         }
+
+        if (attributeName.indexOf('data-') === 0) {
```
Comment (christophehurpeau) [inline src/HTMLElement.js]: you can use startsWith

File: src/utils/DOMStringMap.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
+/**
+ * Class provides DOMStringMap implementation for processing dataset properties.
+ *
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/dataset
+ * @extends {Map}
+ */
+class DOMStringMap {
+    constructor(element) {
```
Comment (christophehurpeau) [inline src/utils/DOMStringMap.js]: also provide jscdoc for the constructor :)

File: src/utils/DOMStringMap.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    return Proxy.create(DOMStringMap.getHandler(list));
+}
+
+/**
+ * @param {string} name
+ * @returns {string}
+ * @private
+ */
+function attributeNameToProperty(name) {
+    return name.replace(/^data-/i, '');
```
Comment (christophehurpeau) [inline src/utils/DOMStringMap.js]: `name.substring('data-'.length)` might be faster ?

File: tests/src/HTMLElement.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    element.dataset.value = 5;
+    assert.equal(element.dataset.value, 5);
+});
+
+test('Element dataset markup using', () => {
+    const document = new Document();
+    document.body.innerHTML = '<b data-attr="some value"></b>';
+    let element = document.body.firstChild;
+    assert.equal(element.dataset.attr, 'some value');
+});
```
Comment (christophehurpeau) [inline tests/src/HTMLElement.js]: It might be great to test every methods of DOMStringMap too

**Comment by christophehurpeau** [comment]
Thanks ! But the coverage decreased `Coverage decreased (-1.3%) to 81.683%`

File: src/utils/DOMStringMap.js (hunk 4 above)
Comment (Indomian) [inline src/utils/DOMStringMap.js]: This will be changed later. Now this class is not used at all.

**Comment by Indomian** [comment]
Completly rewrote everything. Removed unavailable code (DOMStringMap). Added tests. Rebased over current master.

File: lib/HTMLElement.js [hunk 6]
```diff
@@ -59,6 +61,8 @@ var HTMLElement = (function (_Element) {
          * @readonly
          */
         this.classList = new _HTMLElementClassList2['default'](this);
+
+        this._dataset = {};
```
Comment (christophehurpeau) [inline lib/HTMLElement.js]: You can use Map instead of a Object

File: src/HTMLElement.js [hunk 7]
```diff
@@ -25,6 +26,8 @@ export default class HTMLElement extends Element {
          * @readonly
          */
         this.classList = new ClassList(this);
+
+        this._dataset = {};
```
Comment (christophehurpeau) [inline src/HTMLElement.js]: You can use Map instead of a Object

File: src/HTMLElement.js [hunk 8]
```diff
@@ -63,4 +70,7 @@ export default class HTMLElement extends Element {
         }, '') + '>' + (voidElements.indexOf(this.nodeName) !== -1 ? '' : this.innerHTML + '</' + this.nodeName + '>');
     }
 
+    get dataset() {
+        return this._dataset;
```
Comment (christophehurpeau) [inline src/HTMLElement.js]: well... might be not that easy with Map

File: src/HTMLElement.js (hunk 7 above)
Comment (christophehurpeau) [inline src/HTMLElement.js]: maybe not

File: src/utils/escapeAttributeName.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
+/**
+ * Function converts attribute name "data-some-value" to "someValue".
+ *
+ * @param {string} name
+ * @returns {string}
+ */
+function attributeNameToProperty(name) {
```
Comment (christophehurpeau) [inline src/utils/escapeAttributeName.js]: export function

File: src/utils/escapeAttributeName.js [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+    return result.replace(/-(\w)/g, (match, letter) => letter.toUpperCase());
+}
+
+/**
+ * Function converts dataset property name "someValue" to "data-some-value".
+ *
+ * @param {string} name
+ * @returns {string}
+ */
+function propertyNameToAttribute(name) {
```
Comment (christophehurpeau) [inline src/utils/escapeAttributeName.js]: export function

File: src/HTMLElement.js (hunk 7 above)
Comment (Indomian) [inline src/HTMLElement.js]: Map variant was with DOMStringMap. I implemented it using Proxy, but it's not available now in nodejs without --harmony_proxy. Also there is no good shim for that for now.

File: src/utils/escapeAttributeName.js (hunk 10 above)
Comment (Indomian) [inline src/utils/escapeAttributeName.js]: Fixed

File: src/utils/escapeAttributeName.js (hunk 9 above)
Comment (Indomian) [inline src/utils/escapeAttributeName.js]: Fixed

**Comment by Indomian** [comment]
Rebased over current master.

**Pull request merged**